lib/segment/src/index/field_index/immutable_point_to_values.rs

```diff
@@ -0,0 +1,101 @@
+use std::ops::Range;
+
+use common::types::PointOffsetType;
+
+// Flatten points-to-values map
+// It's an analogue of `Vec<Vec<N>>` but more RAM efficient because it stores values in a single Vec.
+// This structure doesn't support adding new values, only removing.
+// It's used in immutable field indices like `ImmutableMapIndex`, `ImmutableNumericIndex`, etc to store points-to-values map.
+#[derive(Debug, Clone, Default)]
+pub struct ImmutablePointToValues<N: Default> {
+    // ranges in `point_to_values_container` which contains values for each point
+    // `u32` is used instead of `usize` because it's more RAM efficient
+    // We can expect that we will never have more than 4 billion values per segment
+    point_to_values: Vec<Range<u32>>,
+    // flattened values
+    point_to_values_container: Vec<N>,
+}
+
+impl<N: Default> ImmutablePointToValues<N> {
+    pub fn new(src: Vec<Vec<N>>) -> Self {
+        let mut point_to_values: Vec<_> = Default::default();
+        let mut point_to_values_container: Vec<_> = Default::default();
```

Comment:
```suggestion
        let mut point_to_values: Vec<_> = Vec::with_capacity(src.len());
        let mut point_to_values_container: Vec<_> = Vec::with_capacity(src.len());
```

The second capacity isn't accurate of course, bit it's cheap and would reduce reallocations by `log2(src.len())`.

We could count the total number of points or estimate based on an average, but that extra work probably isn't worth it.

Reply:
Fixed